### PR TITLE
`schemadiff`: fix index/foreign-key apply scenario

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Fixes https://github.com/vitessio/vitess/issues/16697 by doing two iteration over the `ALTER TABLE` options in `apply()`. Most validations take place on the first iteration. The (newly introduced) second iteration is now able to validate changed after applying all the options.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/16697

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Further changes in this PR: `schemadiff` now detects scenarios where an index is renamed, ie. an index is dropped, and a new index is added in its place with the exact same definition. In such cases, `schemadiff` replaces the `ADD+DROP` clauses with one `RENAME INDEX` clause.

File: go/vt/schemadiff/table.go [hunk 1]
```diff
@@ -2197,20 +2250,19 @@ func (c *CreateTableEntity) apply(diff *AlterTableEntityDiff) error {
...
+			for _, index := range c.TableSpec.Indexes {
+				if strings.EqualFold(index.Info.Name.String(), newKeyName) {
+					return &ApplyDuplicateKeyError{Table: c.Name(), Key: newKeyName}
+				}
+			}
+			for _, index := range c.TableSpec.Indexes {
+				if index.Info.Name.String() == opt.OldName.String() {
+					index.Info.Name = opt.NewName
 				}
 			}
```
Comment (dbussink) [inline go/vt/schemadiff/table.go:2265]: Do we need to error here if the old name doesn't exist?

Comment (shlomi-noach) [reply go/vt/schemadiff/table.go:2268]: Yes, done!

**Comment by dbussink** [comment]
Fix for the CI errors here is in https://github.com/vitessio/vitess/pull/16702

**Review by GuptaManan100 — APPROVED** [review]

**Pull request merged**